src/main/java/com/conveyal/datatools/manager/utils/MergeFeedUtils.java

```diff
@@ -0,0 +1,225 @@
+import static com.conveyal.datatools.manager.models.FeedRetrievalMethod.REGIONAL_MERGE;
+import static com.conveyal.datatools.manager.models.FeedRetrievalMethod.SERVICE_PERIOD_MERGE;
+import static com.conveyal.gtfs.loader.DateField.GTFS_DATE_FORMATTER;
+import static com.conveyal.gtfs.loader.Field.getFieldIndex;
+
+public class MergeFeedUtils {
+    private static final Logger LOG = LoggerFactory.getLogger(MergeFeedUtils.class);
+
+    /**
+     * Get the ids (e.g., trip_id, service_id) for the provided table from the zipfile.
+     */
+    public static Set<String> getIdsForTable(ZipFile zipFile, Table table) throws IOException {
+        Set<String> ids = new HashSet<>();
+        String keyField = table.getKeyFieldName();
+        CsvReader csvReader = table.getCsvReader(zipFile, null);
+        if (csvReader == null) {
+            LOG.warn("Table {} not found in zip file: {}", table.name, zipFile.getName());
+            return ids;
+        }
+        Field[] fieldsFoundInZip = table.getFieldsFromFieldHeaders(csvReader.getHeaders(), null);
+        // Get the key field (id value) for each row.
+        int keyFieldIndex = getFieldIndex(fieldsFoundInZip, keyField);
+        while (csvReader.readRecord()) ids.add(csvReader.get(keyFieldIndex));
+        csvReader.close();
+        return ids;
+    }
+
+    /**
+     * Construct stop_code failure message for {@link com.conveyal.datatools.manager.jobs.MergeFeedsJob} in the case of
+     * incomplete stop_code values for all records.
+     */
+    public static String stopCodeFailureMessage(int stopsMissingStopCodeCount, int stopsCount, int specialStopsCount) {
+        return String.format(
+            "If stop_code is provided for some stops (for those with location_type = " +
+                "empty or 0), all stops must have stop_code values. The merge process " +
+                "found %d of %d total stops that were incorrectly missing stop_code values. " +
+                "Note: \"special\" stops with location_type > 0 need not specify this value " +
+                "(%d special stops found in feed).",
+            stopsMissingStopCodeCount,
+            stopsCount,
+            specialStopsCount
+        );
+    }
+
+    /**
+     * Collect zipFiles for each feed version before merging tables.
+     * Note: feed versions are sorted by first calendar date so that future dataset is iterated over first. This is
+     * required for the MTC merge strategy which prefers entities from the future dataset over active feed entities.
+     */
+    public static List<FeedToMerge> collectAndSortFeeds(Set<FeedVersion> feedVersions) {
+        return feedVersions.stream()
+            .map(version -> {
+                try {
+                    return new FeedToMerge(version);
+                } catch (Exception e) {
+                    LOG.error("Could not create zip file for version: {}", version.version);
+                    return null;
```

Comment:
This probably deserves a report to Bugsnag.

Reply:
Handled in https://github.com/ibi-group/datatools-server/pull/376/commits/8179f873f950f83394f7c628131f9b2e73fc72e1.